### PR TITLE
Autocorrect Input using spellchecker library. #11 

### PR DESCRIPTION
**Pull request by itay-goldraich**
Rewrote the _pre_process function.
Added the _autocorrect function.
Added a data dir.
Added a drug_names.txt file.
Added an autocorrect test to the test file.
#11

### CODE REVIEW

File: parsigs/parse_sig_api.py [hunk 1]
```diff
@@ -86,27 +87,39 @@ def _parse_sig(sig: str, model: Language):
 
     return _create_structured_sigs(model_output)
 
+def _autocorrect(sig):
+    spell = SpellChecker()
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:91]: spell checker should be defined once globally

File: parsigs/parse_sig_api.py [hunk 2]
```diff
@@ -86,27 +87,39 @@ def _parse_sig(sig: str, model: Language):
...
+    spell.word_frequency.load_text_file(os.path.join('data', 'drug_names.txt'))
+    corrected_words = [spell.correction(word) if not any(char.isdigit() for char in word) else word for word in sig.split()]
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:93]: this is too much code in one line :) tru seperating the logic including function definitions to improve readability

File: parsigs/parse_sig_api.py [hunk 3]
```diff
@@ -86,27 +87,39 @@ def _parse_sig(sig: str, model: Language):
...
-            output_words.append(word)
-    sig = ' '.join(output_words)
 
+def _pre_process(sig):
+    #text to lower case and remove leading and ending whitespace
+    sig = sig.lower().strip()
+    # Spell check
+    sig = _autocorrect(sig)
+    #manual replacements of familiar phrases and characters
+    replacements = {
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:104]: This does not contain the previoud logic with the regex, also why there are spaces before and after the words? 
I think it is better using the old logic, or recreating it in a dictionary similar to what you did but globally as a constant

File: tests/test_parsig_api.py [hunk 4]
```diff
@@ -55,8 +55,8 @@ def test_parse_sig_short_form(self):
         self.assertEqual(result, expected)
 
     def test_parse_sig_no_form(self):
-        sig = "Take 1 codaine 3 times a day"
-        expected = StructuredSig(drug='codaine', form=None, strength=None, frequencyType="Day", interval=3, singleDosageAmount=1.0, periodType=None, periodAmount=None, takeAsNeeded=False)
+        sig = "Take 1 codeine 3 times a day"
```
Comment (royashcenazi) [inline tests/test_parsig_api.py:58]: Good catch :)

File: tests/test_parsig_api.py [hunk 5]
```diff
@@ -102,3 +102,9 @@ def test_parse_multiple_instructions3(self):
         result = self.sig_parser.parse(sig)
         expected = [first_expected, second_expected, third_expected]
         self.assertEqual(result, expected)
+        
+    def test_autocorrect(self):
+        sig = "Take 1 talbet 3 tiems a day for 2 wekes"
```
Comment (royashcenazi) [inline tests/test_parsig_api.py:107]: Add at least 5 more tests please :)

**Comment by royashcenazi** [comment]
@itay-goldraich thanks very nice idea!
Additional to my comments on the code, you also need to add to the `requirements.txt` and `pyproject.toml` the new dependancy

**Review by royashcenazi — CHANGES_REQUESTED** [review]

File: tests/test_parsig_api.py (hunk 4 above)
Comment (itay-goldraich) [reply tests/test_parsig_api.py:58]: Thanks

File: parsigs/parse_sig_api.py (hunk 3 above)
Comment (itay-goldraich) [reply parsigs/parse_sig_api.py:104]: I've changed the replacements dictionary to preserve the same desired outcome.

File: parsigs/parse_sig_api.py (hunk 1 above)
Comment (itay-goldraich) [reply parsigs/parse_sig_api.py:91]: Done.

File: parsigs/parse_sig_api.py (hunk 2 above)
Comment (itay-goldraich) [reply parsigs/parse_sig_api.py:93]: Fixed.

File: tests/test_parsig_api.py (hunk 5 above)
Comment (itay-goldraich) [reply tests/test_parsig_api.py:107]: There are 7 tests.
I think there might be a better way to test the autocorrect feature, but I will revisit the test issue in the future as the code passed all of the current tests.

**Comment by royashcenazi** [comment]
Dont see your changes yet - seems like you did not push them

File: parsigs/parse_sig_api.py [hunk 6]
```diff
@@ -36,6 +37,21 @@
 takeAsNeeded : bool
     Some instructions contains a statement that the medication should be taken as needed by patient
 """
+# Global Variables
+spell = SpellChecker()
+spell.word_frequency.load_text_file(os.path.join('data', 'drug_names.txt'))
+replacements = {
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:43]: please remove this and return to the previous logic :)

File: pyproject.toml [hunk 7]
```diff
@@ -15,6 +15,7 @@ dependencies = [
     "spacy>=3.0,<4.0",
     "word2number",
 #    "en-parsigs @ https://huggingface.co/royashcenazi/en_parsigs/resolve/main/en_parsigs-any-py3-none-any.whl",
+    "spellchecker<= 0.5.5",
```
Comment (royashcenazi) [inline pyproject.toml:18]: can you send the github link to this dependancy? 
We need to make sure there are no network (api calls) in order to keep patient data secured
other than that looks good!

Comment (itay-goldraich) [reply pyproject.toml:18]: The link is in the description of the issue, I can send it in PR if you wish.

Comment (royashcenazi) [reply pyproject.toml:18]: @itay-goldraich seems the dependancy than should be `pyspellchecker` and not `spellchecker` no? 
Fix this, run tests and we can merge :)

Comment (itay-goldraich) [reply pyproject.toml]: Fixed.

File: parsigs/parse_sig_api.py [hunk 8]
```diff
@@ -87,26 +88,39 @@ def _parse_sig(sig: str, model: Language):
...
+            corrected_words.append(corrected_word)
+    sig = ' '.join(corrected_words)
+    return sig
 
-    sig = _add_space_around_parentheses(sig)
 
+def _pre_process(sig):
+    sig = _autocorrect(sig)
+    replacements = {
+            '@': 'a',
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:109]: @itay-goldraich you have not returned this to previous logic, please do so we can merge

File: parsigs/parse_sig_api.py [hunk 9]
```diff
@@ -87,26 +88,39 @@ def _parse_sig(sig: str, model: Language):
...
 
-def _pre_process(sig):
-    sig = sig.lower().replace('twice', '2 times').replace("once", '1 time').replace("nightly", "every night")
+def _autocorrect(sig):
+    spell = SpellChecker()
+    spell.word_frequency.load_text_file(os.path.join('data', 'drug_names.txt'))
+    sig = sig.lower().strip()
+    corrected_words = []
+    for word in sig.split():
+        if not re.match(r"^[a-zA-Z]+$",word) or spell.known([word]):
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:97]: define the REGEX as a global constant with a meaningful name and use it here

File: parsigs/parse_sig_api.py [hunk 10]
```diff
@@ -1,12 +1,13 @@
...
-
 import spacy
 from word2number import w2n
 from dataclasses import dataclass
 import re
 from spacy import Language
 from itertools import chain
 import copy
+import os
+from spellchecker import SpellChecker
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:10]: something does not make sense, if the dependency is pyspellchecker than the import should be the same

File: parsigs/parse_sig_api.py (hunk 9 above)
Comment (itay-goldraich) [reply parsigs/parse_sig_api.py:97]: After a conversation on the project's Discord, I'll write a short record for this regex.

File: parsigs/parse_sig_api.py [hunk 11]
```diff
@@ -87,14 +88,27 @@ def _parse_sig(sig: str, model: Language):
     return _create_structured_sigs(model_output)
 
 
-def _pre_process(sig):
-    sig = sig.lower().replace('twice', '2 times').replace("once", '1 time').replace("nightly", "every night")
+def _autocorrect(sig):
+    spell = SpellChecker()
+    spell.word_frequency.load_text_file(os.path.join('data', 'drug_names.txt'))
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:93]: The spellchecker and file loading should be defined as global variables not to be loaded every call to the method

Regarding the added file- please add it to a directory called `resources`

**Review by royashcenazi — APPROVED** [review]

**Pull request merged**